lib/SIL/SILOpenedArchetypesTracker.cpp

```diff
@@ -0,0 +1,129 @@
+// Can be used to incrementally populate the mapping, e.g.
+// if it is done when performing a scan of all instructions
+// inside a function.
+void SILOpenedArchetypesTracker::registerOpenedArchetypes(
+    const SILInstruction *I) {
+  assert((!I->getParent() || I->getFunction() == &F) &&
+         "Instruction does not belong to a proper SILFunction");
+  auto Archetype = getOpenedArchetypeOf(I);
+  if (Archetype)
+    addOpenedArchetypeDef(Archetype, I);
+}
+
+// Register opened archetypes whose definitions are referenced by
+// the typedef operands of this instruction.
+void SILOpenedArchetypesTracker::registerUsedOpenedArchetypes(
+    const SILInstruction *I) {
+  assert((!I->getParent() || I->getFunction() == &F) &&
+         "Instruction does not belong to a proper SILFunction");
+  for (auto &Op : I->getOpenedArchetypesOperands()) {
+    auto TypeDef = Op.get();
+    assert(isa<SILInstruction>(TypeDef) &&
+           "typedef operand should refer to a SILInstruction");
+    addOpenedArchetypeDef(getOpenedArchetypeOf(cast<SILInstruction>(TypeDef)),
+                          TypeDef);
+  }
+}
+
+// Unregister archetypes opened by a given instruction.
+// Should be only called when this instruction is to be removed.
+void SILOpenedArchetypesTracker::unregisterOpenedArchetypes(
+    const SILInstruction *I) {
+  assert(I->getFunction() == &F &&
+         "Instruction does not belong to a proper SILFunction");
+  auto Archetype = getOpenedArchetypeOf(I);
+  if (Archetype)
+    removeOpenedArchetypeDef(Archetype, I);
+}
+
+void SILOpenedArchetypesTracker::handleDeleteNotification(
+    swift::ValueBase *Value) {
+  if (auto I = dyn_cast<SILInstruction>(Value))
+    if (I->getFunction() == &F)
+      unregisterOpenedArchetypes(I);
+}
+
+/// Find an opened archetype defined by an instruction.
+/// \returns The found archetype or empty type otherwise.
+CanType swift::getOpenedArchetypeOf(const SILInstruction *I) {
+  if (isa<OpenExistentialAddrInst>(I) || isa<OpenExistentialRefInst>(I) ||
+      isa<OpenExistentialBoxInst>(I) || isa<OpenExistentialMetatypeInst>(I)) {
+    auto Ty = getOpenedArchetypeOf(I->getType().getSwiftRValueType());
+    assert(Ty->isOpenedExistential() && "Type should be an opened archetype");
+    return Ty;
+  }
+
+  return CanType();
+}
+
+CanType swift::getOpenedArchetypeOf(CanType Ty) {
+  if (!Ty)
+    return CanType();
+  while (auto MetaTy = dyn_cast<AnyMetatypeType>(Ty)) {
+    Ty = MetaTy->getInstanceType().getCanonicalTypeOrNull();
+  }
+  if (Ty->isOpenedExistential())
+    return Ty;
+  return CanType();
+}
+
+/// Return the defintion of an opened archetype.
+/// It it is an opened archetype, but its definition is not known
+/// yet, return a SILUndef with the type of this opened archetype
+/// Otehrwise return a SILUndef of non-archetype type.
+SILValue
+swift::getOpenedArchetypeDef(SILType SILTy, SILModule &M,
+                             SILOpenedArchetypesTracker *OpenedArchetypes) {
```

Comment:
Is this function still needed, along with UndefAllocator, getUndefValue, getEmptyTypeDef, and so on?